### ISSUE TITLE
problem dependency with trust-dns-proto

### ISSUE THREAD

**Issue by Stargateur** [actix/actix#345]
I'm using mongodb 9.0.1 that use:

```toml
trust-dns-proto = "0.18.0"
trust-dns-resolver = "0.18.0"
```
https://github.com/mongodb/mongo-rust-driver/blob/bfd69496baf56ba89668f589dfed8733aa24c42d/Cargo.toml#L35

And actix 0.9.0 that use:
```rust
trust-dns-proto = { version = "=0.18.0-alpha.2", optional = true, default-features = false }
trust-dns-resolver = { version = "=0.18.0-alpha.2", optional = true, default-features = false }
```
https://github.com/actix/actix/blob/5eed1043438b9b5d62fec103db3665657f9c61bc/Cargo.toml#L58

It's produce the following error message:

```none
error: failed to select a version for `trust-dns-proto`.
    ... required by package `mongodb v0.9.1`
    ... which is depended on by `blender v1.0.5 (/home/stargateur/blender)`
versions that meet the requirements `^0.18.0` are: 0.18.1, 0.18.0

all possible versions conflict with previously selected packages.

  previously selected package `trust-dns-proto v0.18.0-alpha.2`
    ... which is depended on by `actix v0.9.0`
    ... which is depended on by `blender v1.0.5 (/home/stargateur/blender)`

failed to select a version for `trust-dns-proto` which could resolve this conflict
```

**Comment by JohnTitor** [comment]
We're aware of that and in roadmap of 0.10 :)

**Comment by Stargateur** [comment]
@JohnTitor I'm not good with semver tag but shouldn't actix 9.0 be 0.9.0-alpha.3 ?

**Comment by JohnTitor** [comment]
> I'm not good with semver tag but shouldn't actix 9.0 be 0.9.0-alpha.3 ?

Uhm? Where do you refer to?

**Comment by Stargateur** [comment]
actix versioning https://crates.io/crates/actix/versions.

I don't understand why actix release a 9.0 not compatible with other crates that use non alpha release of trust-dns-proto. I mean I just can't compile with 9.0 at this point, so 9.0 is unusable.

**Comment by JohnTitor** [comment]
> I don't understand why actix release a 9.0 not compatible with other crates that use non alpha release of trust-dns-proto.

I believe you mean 0.9.0, not 9.0 :) Yeah, it's not ideal that 0.9.0 relies on their pinned versions.

> I mean I just can't compile with 9.0 at this point, so 9.0 is unusable.

0.9.0 is still usable, however, when you can avoid conflicts. I'd like to make a new alpha-release, i.e. 0.10.0 (alpha) with updating their deps versions since it'll also break backward compatibility. Unfortunately, I really don't have bandwidth enough now so the work will be late :/ Any contributions are welcome!

**Comment by Stargateur** [comment]
Well, I struggle to just trying to use/understand actix so I will not be very effective :p but I will clearly keep that in mind.

mongodb issues I found:

 - https://jira.mongodb.org/browse/RUST-287
 - https://jira.mongodb.org/browse/RUST-290

Their will update to trust-dns-proto 0.19, but the problem come from actix so... that will happen again. (well, even if I think cargo is also the problem... semver is strange).

**Comment by JohnTitor** [comment]
The conflict occurs between 0.18.0 and 0.18.0-alpha.2 so indeed using mongodb 0.9.0 is an avoidance.

You can try:
```toml
[dependencies]
actix = "0.9.0"
mongodb = "=0.9.0"
```

**Comment by Stargateur** [comment]
> No worries, but be on the lookout for this fix, because 0.9.0 uses old trust-dns-* versions that still rely on tokio 0.1 (so dependency bloat) [link](https://jira.mongodb.org/browse/RUST-287?focusedCommentId=2881059&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2881059)

Well, that will do it for me for the moment, thanks for the tips, I let you choice when to close this issue.

**Comment by Jonathas-Conceicao** [comment]
I have started a migration that does compile, but couldn't test it properly so far. If anyone can help testing and let me know of any problems would be of great help.

My migration attempt is available on [a branch on my fork](https://github.com/Jonathas-Conceicao/actix/tree/topic/upgrade_trust_dns).

One important diference from the older implementation to the one I currently have is: the old implementation would use the Actor's context as a "background runner" to resolve spawned by `trust_dns_resolver::AsyncResolver`, the new API makes it a little hard to reproduce the same behavior, so I decided to let it spawn it's futures in to the current runtime directly since it's much simple, however I'm not sure if this is a problem or not. If this isn't desirable for some reason, I think the `Addr<Resolver>` can be made into Runtime for the `trust_dns_resolver::AsyncResolver`, but it would add some considerable complexity and higher load of message passing.
